### PR TITLE
Synced flush backport

### PR DESCRIPTION
**Pull request by brwe**
Was relatively unproblematic, only SyncedFlushService was a little hairy

### CODE REVIEW

File: src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java [hunk 1]
```diff
@@ -358,7 +410,7 @@ public void onResponse(PreSyncedFlushResponse response) {
 
                 @Override
                 public void onFailure(Throwable e) {
-                    logger.trace("{} error while performing pre synced flush on [{}], skipping", e, shardId, shard);
+                    logger.trace("{} error while performing pre synced flush on [{}], skipping", shardId, e, shard);
```
Comment (spinscale) [inline src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java]: this looks flipped and does not log the exception?

File: src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 2]
```diff
@@ -178,6 +178,8 @@ public void phase1(final SnapshotIndexCommit snapshot) throws ElasticsearchExcep
                 }
                 // we shortcut recovery here because we have nothing to copy. but we must still start the engine on the target.
                 // so we don't return here
+                logger.trace("[{}][{}] skipping [phase1] to {} - identical sync id [{}] found on both source and target", indexName, shardId,
```
Comment (spinscale) [inline src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: missing `[]` around targetNode?

File: src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java [hunk 3]
```diff
@@ -252,41 +315,24 @@ void sendSyncRequests(final String syncId, final List<ShardRouting> shards, Clus
...
             }
             final Engine.CommitId expectedCommitId = expectedCommitIds.get(shard.currentNodeId());
             if (expectedCommitId == null) {
                 logger.trace("{} can't resolve expected commit id for node {}, skipping for sync id [{}]. shard routing {}", shardId, shard.currentNodeId(), syncId, shard);
                 results.put(shard, new SyncedFlushResponse("no commit id from pre-sync flush"));
-                contDownAndSendResponseIfDone(syncId, shards, shardId, listener, countDown, results);
+                contDownAndSendResponseIfDone(syncId, shards, shardId, totalShards, listener, countDown, results);
                 continue;
             }
             assert node.version().onOrAfter(Version.V_1_6_0) : " node with version " + node.version() + " can not have a sync commit ID";
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java]: not part of this change but I just noticed it - should be " can not have an *expected* commit ID " (not sync)

File: src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java [hunk 4]
```diff
@@ -252,41 +315,24 @@ void sendSyncRequests(final String syncId, final List<ShardRouting> shards, Clus
...
-                }
-            };
             if (state.nodes().localNode().id().equals(node.id())) {
-                execute(ThreadPool.Names.FLUSH, new Callable<SyncedFlushResponse>() {
-                    @Override
-                    public SyncedFlushResponse call() throws Exception {
-                        return performSyncedFlush(new SyncedFlushRequest(shard.shardId(), syncId, expectedCommitId));
-                    }
-                }, currentShardListener);
+                SyncedFlushResponse existing = results.put(shard, performSyncedFlush(new SyncedFlushRequest(shard.shardId(), syncId, expectedCommitId)));
```
Comment (bleskes) [inline src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java]: why did you remove the FLUSH executor?

File: src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java (hunk 1 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java]: ++

File: src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 2 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Disco Nodes toString adds []

File: src/test/java/org/elasticsearch/gateway/local/RecoveryBackwardsCompatibilityTests.java [hunk 5]
```diff
@@ -77,7 +77,8 @@ public void testReusePeerRecovery() throws Exception {
         indexRandom(true, builders);
         ensureGreen();
         if (randomBoolean()) { // just make sure it doesn't break anything - we seal before we actually bump replicas
-            backwardsCluster().internalCluster().client().admin().indices().prepareSealIndices("test").get();
+            logger.info("--> trying to sync flush");
+            assertEquals(SyncedFlushUtil.attemptSyncedFlush(internalCluster(), "test").failedShards(), 0);
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/gateway/local/RecoveryBackwardsCompatibilityTests.java]: don't we need backwardsCluster().internalCluster() here? also, are we sure we have at least one local (client) node? o.w. there is not SyncedFlushtService instance..

**Comment by bleskes** [comment]
Thx @brwe for picking this up. Left some minor comments

File: src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 2 above)
Comment (spinscale) [inline src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: thx for the clarification

**Comment by spinscale** [comment]
did not spot any huge differences on my side, except the same question @bleskes had about removing the FLUSH executor...

apart from that LGTM

File: src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java (hunk 4 above)
Comment (brwe) [inline src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java]: I thought it was not needed anymore but think now it blocks and might not even be in the right thread pool...will add again. thanks also @spinscale

File: src/test/java/org/elasticsearch/gateway/local/RecoveryBackwardsCompatibilityTests.java (hunk 5 above)
Comment (brwe) [inline src/test/java/org/elasticsearch/gateway/local/RecoveryBackwardsCompatibilityTests.java]: yes, should be backwardsCluster()... we always have current version nodes in the test unless annotated otherwise. is that what you mean?

Comment (bleskes) [inline src/test/java/org/elasticsearch/gateway/local/RecoveryBackwardsCompatibilityTests.java]: Yeah, that is what I meant. I guess we should check and protect for having no current version nodes, no?

**Comment by brwe** [comment]
ok, all done. I let the bwc tests run and found that there we also have to check if it is a bwc test or not when we call indexRandom(). pushed a fix for that.

File: src/test/java/org/elasticsearch/gateway/local/RecoveryBackwardsCompatibilityTests.java (hunk 5 above)
Comment (brwe) [inline src/test/java/org/elasticsearch/gateway/local/RecoveryBackwardsCompatibilityTests.java]: I will add that, also to indexRandom() in ElasticsearchIntegrationTest. However, after running the test I found that the actual problem is that we check for no failures here - which is not good because the check is doomed to fail if a shard was an older version node. will remove the assert as well.

**Comment by brwe** [comment]
fixed the bwc test and added checks to make sure we actually have at least one current version node when we run the bwc tests. want to take another look?

File: src/test/java/org/elasticsearch/test/ElasticsearchIntegrationTest.java [hunk 6]
```diff
@@ -1480,8 +1478,22 @@ private void postIndexAsyncActions(String[] indices, List<CountDownLatch> inFlig
...
                             new LatchedActionListener<FlushResponse>(newLatch(inFlightAsyncOperations)));
                 } else {
-                    client().admin().indices().prepareSealIndices(indices).execute(
-                            new LatchedActionListener<SealIndicesResponse>(newLatch(inFlightAsyncOperations)));
+                    InternalTestCluster internalTestCluster = null;
+                    if (!isInternalCluster()) {
+                        // bwc test
+                        if (cluster() instanceof CompositeTestCluster) {
+                            CompositeTestCluster compositeTestCluster = (CompositeTestCluster) cluster();
+                            if (compositeTestCluster.numNewDataNodes() > 0) {
```
Comment (bleskes) [inline src/test/java/org/elasticsearch/test/ElasticsearchIntegrationTest.java]: can we add a comment as to why this is needed here? I'm not sure it's trivial to figure out.

**Comment by bleskes** [comment]
LGTM. Left one little comment. Feel free to push.
